### PR TITLE
Implement coherence delta vs. date plot

### PR DESCRIPTION
**Pull request by nackerley**
### Implements
This pull request implements:
- #4 
- #8 

## Outstanding
- Deprecate global variables:
> - I thought I saw a global variable used, but in fact this is only the case the slider and intensity plots are un-commented. That, said there were "redefinition of global variable" warnings which I've scrubbed out.
> - At a certain point I was thinking that when we get to the point of being able to switch between rasters as described in #3 and between polygon overlays as per #2 it may be necessary to record the state of the relevant radio buttons to determine what to display, when different points are selected on the coherence plot, and that the way to do this is using dcc.Store. But now I'm thinking it's probably possible to query the state of the radio buttons directly.
> - Conclusion: I don't foresee any need to share data between components.
- #11
- #12

### CODE REVIEW

**Comment by nackerley** [comment]
I see that some checks were not successful, but I'm pretty sure the errors are all in the setup, not in the code I've pushed.

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,3 @@
 virtualenv/*
-.vscode/settings.json
-config.ini
```
Comment (drotheram) [inline .gitignore:3]: Config.ini should remain in the .gitignore file

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
```
Comment (drotheram) [inline .vscode/launch.json:4]: Can we remove the IntelliSense plug?

Comment (nackerley) [reply .vscode/launch.json:4]: Absolutely!

File: .gitignore (hunk 1 above)
Comment (nackerley) [reply .gitignore:3]: Why? I don't understand.

My motivation in including it was that the system didn't work "out of the box" for lack of a config.ini, yet there was a sample_config.ini.

**Comment by nackerley** [comment]
Any progress on getting pylint to pass?

File: app/dash_app.py [hunk 3]
```diff
@@ -1,383 +1,397 @@
...
-    if not clickData:
-        return 'cite:20210717_HH_20210903_HH.adf.wrp.geo'
-    dates = '{}_HH_{}_HH'.format(json.dumps(clickData['points'][0]['x'],
-                                            indent=2),
-                                 json.dumps(clickData['points'][0]['y'],
-                                            indent=2))
-    dates = dates.replace('-', '').replace('"', '')
-    print(dates)
-    return 'cite:{}.adf.wrp.geo'.format(dates)
+        # html.Div(
```
Comment (drotheram) [inline app/dash_app.py:274]: Let's go ahead and remove this commented code. It is in the git history if we ever need it

File: app/dash_app.py [hunk 4]
```diff
@@ -1,383 +1,397 @@
...
+    for group, subset in coh_wide.groupby(groups):
+        row = int(groups.nunique() - group)
+        fig.add_trace(
+            Heatmap(
+                z=subset.values,
+                x=subset.columns,
+                y=subset.index,
+                hovertemplate=(
+                    'Second: %{x}<br>'
+                    'First: -%{y} days<br>'
```
Comment (drotheram) [inline app/dash_app.py:170]: Would like to see both dates on the hover in addition to time between dates. This is useful but there are many scenarios where the analyst will want to know if the interferogram displayed covers a specific event or date. Datetime module should be able to calculate this easily with what's already here. Let's keep the temporal baseline as an additional piece of info. The integer doesn't need to be negative.

File: app/dash_app.py [hunk 5]
```diff
@@ -1,383 +1,397 @@
...
+#     date = date.replace('-', '')
+#     layers = f'cite:{date}_HH.rmli.geo.db'
+#     print(f'Updating backscatter: {layers}')
+#     return layers
 
 
 if __name__ == '__main__':
-    app.run_server(host='0.0.0.0', port=8050, debug=False)
+    # TODO login and set up - or at least test - port forwarding
+    # See https://shorturl.at/lnSY1
```
Comment (drotheram) [inline app/dash_app.py:396]: Remove ToDo comments and create repo issue

File: app/dash_app.py [hunk 6]
```diff
@@ -1,383 +1,397 @@
...
+    """Parse configuration from supplied file."""
+    config_obj = configparser.ConfigParser()
+    config_obj.read(args)
+    return config_obj
+
+
+config = get_config_params('config.ini')
+GEOSERVER_ENDPOINT = config.get('geoserver', 'geoserverEndpoint')
+
+# TODO add support for some or all of the following parameters to config
```
Comment (drotheram) [inline app/dash_app.py:39]: remove 'ToDo' and create issue

File: app/dash_app.py [hunk 7]
```diff
@@ -1,383 +1,397 @@
...
+    'Tiles courtesy of the '
+    '<a href="https://usgs.gov/">U.S. Geological Survey</a>')
+BASEMAP_NAME = 'USGS Topo'
+
+# coherence plotting configuration
+GROUP_GAP_DAYS = 60
+CMAP_NAME = 'RdYlBu_r'
+COH_LIMS = (0.25, 0.65)
+
+# TODO read target configuration from database
```
Comment (drotheram) [inline app/dash_app.py:60]: remove todo and create repo issue

File: .gitignore (hunk 1 above)
Comment (drotheram) [reply .gitignore:3]: We should add a note in the Readme for the user to copy `sample_config.ini` to `config.ini` the contents of config.ini are only relevant to users with access to our cloud resources.

**Review by drotheram — CHANGES_REQUESTED** [review]
This is a good pass at adding the coherence timedelta. I like the added functionality of the hover. Before we merge I am requesting some changes. 

- Coherence timedelta plot gaps. This is probably going to be unsustainable for any of our beams that get bumped seasonally each new year will require a new plot gap
- Map and plots are too tall everything should be visible without scrolling. I had to zoom out to 75% zoom for the dashboard not to appear clausterphobic. This is probably because I assigned pixel heights rather than % heights initially. With the added plot below the screen gets too crowded.
- Coherence dot colouring scale is too wide. Anything above 0.3 should have a very hot colour, between 0.2 - 0.3 should still be relatively hot. With this stretch 0.3 coherence value interferograms are shades of blue
- Y-axis of coherence plot should probably be named 'Temporal Baseline (days)' as standard InSAR phraseology

File: .gitignore (hunk 1 above)
Comment (nackerley) [reply .gitignore:3]: Agreed that if we're not going to include `config.ini` in the repo it should documented. But it seems simpler to just include it in the repo - what's the downside?

File: app/dash_app.py (hunk 4 above)
Comment (nackerley) [reply app/dash_app.py:170]: Agreed that showing the date would be better - it's going to take a bit of work to implement that, though. I don't think it's going to be possible to calculate it on the fly, I think you need to supply the extra data using the `customdata` attribute as described at https://plotly.com/python/hover-text-and-formatting/.

The negative sign is to indicate that the first date is always before (less than) the second. I thought it was clearer than just giving the delta.

File: app/dash_app.py (hunk 5 above)
Comment (nackerley) [reply app/dash_app.py:396]: Issue exists #12. We can remove the TODO when the issue is cleared.

File: app/dash_app.py (hunk 6 above)
Comment (nackerley) [reply app/dash_app.py:39]: Issue exists #11 and we can remove TODO when it is closed.

File: .gitignore (hunk 1 above)
Comment (drotheram) [reply .gitignore:3]: Listing our service endpoints, even if secure, could potentially be a security target

**Comment by drotheram** [comment]
Re: Coherence timedelta plot gaps. My suggestion is to limit visible scale to a fixed temporal baseline, such as 60-80 days. and allow the user to zoom in/out, pan, or zoom to the full network. This visualization needs to support the general case of coherence networks over long periods and different sensors with different repeat pass intervals. We will also likely see different seasonal trends if we begin to examine volcano hazards outside of Canada. 

RE: "Remember that part of keeping this usable is going to be "clearing out" date-pairs that turn out to be useless, like summer compared to winter" - Agreed, we can discard interferograms below a certain coherence threshold but once they have been processed and an average coherence value has been generated, we might as well retain that value so that we know empirically that the coherence was calculated and it was low.

File: app/dash_app.py (hunk 7 above)
Comment (nackerley) [reply app/dash_app.py:66]: Issue #11 exists, and TODO can be removed when it is resolved.

File: app/dash_app.py (hunk 4 above)
Comment (nackerley) [reply app/dash_app.py:183]: Added issue #14 to be resolved by a future pull request.

File: .gitignore (hunk 1 above)
Comment (nackerley) [reply .gitignore:3]: Done.

**Comment by nackerley** [comment]
@drotheram, I have fixed the CI/CD linters. We now have a PEP8 check which must pass, and a requirement that the pylint score be high (>9). See https://github.com/Volcano-Risk-Reduction-in-Canada/Volcanic-Interpretation-Workbench/tree/feature/coherence_timedelta/.github/workflows and the green checkmarks above.

I think I've addressed all the changes you've requested.

**Review by drotheram — APPROVED** [review]

**Pull request merged**